### PR TITLE
Secondary auto-reboot support

### PR DESCRIPTION
**Pull request by lbonn**

### CODE REVIEW

File: recipes-test/demo-config/primary-config.bb [hunk 1]
```diff
@@ -11,6 +11,7 @@ PRIMARY_SECONDARIES ?= "${SECONDARY_IP}:${SECONDARY_PORT}"
 SRC_URI = "\
     file://30-secondary-config.toml \
     file://ip_secondary_config.json \
+    ${@('file://' + d.getVar('SOTA_SECONDARY_CONFIG')) if d.getVar('SOTA_SECONDARY_CONFIG') else ''} \
```
Comment (mike-sul) [inline recipes-test/demo-config/primary-config.bb]: I am not sure if I get what is this for?

Comment (lbonn) [reply recipes-test/demo-config/primary-config.bb]: bitbake needs the file in `SRC_URI` to track a dependency.

Without this change, if you build an image then modify the file and rebuild, the new file version won't be taken into account and bitbake will just give you the same image from cache

With this change, the image is rebuilt as expected.

Comment (mikesul-vd) [reply recipes-test/demo-config/primary-config.bb]: ok, thanks for the explanation.

Comment (mike-sul) [reply recipes-test/demo-config/primary-config.bb]: ok, thanks for the explanation!

**Comment by lbonn** [comment]
I've just realized (from @Raigi's report) that the aktualizr-auto-reboot recipe can simply be reused for secondary image and if it's in `IMAGE_INSTALL_append` used for both primary and secondary, it will also take effect on both images.

So maybe I'll just remove this variable and document that properly?

**Comment by mike-sul** [comment]
> I've just realized (from @Raigi's report) that the aktualizr-auto-reboot recipe can simply be reused for secondary image and if it's in `IMAGE_INSTALL_append` used for both primary and secondary, it will also take effect on both images. And the corollary is that if both are used together, it creates conflict during file system creation.
> 
> So maybe I'll just remove this variable and document that properly?

I suppose, having a single way to configure an auto-reboot is more consistent, on the other hand it will require additional manipulations if the value of this parameter should be different on Primary and Secondary.

**Comment by patrickvacek** [comment]
> So maybe I'll just remove this variable and document that properly?

Yeah, that's probably best.

> it will require additional manipulations if the value of this parameter should be different on Primary and Secondary.

True, but we already have that situations for plenty of other configuration variables. An annoying result of using the same build environment for bitbaking images for two different types of build targets.

**Comment by lbonn** [comment]
@mike-sul that's correct.

Just note that if it needs to have different values and be configurable from a single local.conf/site.conf, aktualizr-auto-reboot probably can't be used as is and we'd need to also define a similar variable for the primary image.

But I think that users will generally want to use the same value for both primary and secondary. If they need more flexibility, they can also selectively add aktualizr-auto-reboot to their different images.

I also have the feeling that having many variables to define in local.conf is unidiomatic yocto (for better or worse), compared to using recipes and bbappends.

**Comment by lbonn** [comment]
poky broke today: https://git.yoctoproject.org/cgit/cgit.cgi/poky/commit/?id=8efa405ee0ab9ae0e674fe8bc9551f4367b4c269. There is a fix in master-next, hoping it will come soon.

**Comment by Raigi** [comment]
Did something change in the way image is run in master? I am not able to provision a device anymore when running: 
`/home/user/meta-updater/scripts/run-qemu-ota --no-gui  --dir /home/user/virtual_master/build/tmp/deploy/images/ --machine qemux86-64 --secondary-network primary-image`
I was able to provision a device with the same command from zeus branch.

**Comment by patrickvacek** [comment]
@Raigi what's the error?

**Comment by patrickvacek** [comment]
> Did something change in the way image is run in master?

You wouldn't be trying to provision a Primary that is expecting a Secondary without launching the Secondary, would you? Otherwise, the networking logic in the test images for Primaries and Secondaries did slightly change a few weeks ago, but we backported that to all the supported release branches. When I manually launch a primary-image and secondary-image, I'm having trouble getting them to talk to each other, but I've gotten it to work at least once.

**Comment by patrickvacek** [comment]
Summary of discussion with Raigi:

Both of us can get the Primary and Secondary to start talking and successfully provision only if we stop the systemd services and manually restart them. For some reason, they will not connect without doing that, and in my experience, even that isn't always successful. Raigi reports seeing this with this branch and on meta-updater master but none of the release branches. She's reproduced it with OTF and manually. I've reviewed the recent meta-updater CI runs, and while most of them failed for various reasons over the past few days, the Secondary tests always succeeded.

What could be different between OTF and the meta-updater CI? Almost nothing has been merged to master that wasn't already backported to the release branches. @mike-sul and @lbonn can you think of anything that might have changed in other layers or anything else that might be thwarting this?

**Comment by mike-sul** [comment]
> . Raigi reports seeing this with this branch and on meta-updater master but none of the release branches.

So, can I try the master branch of meta-updater to reproduce it?

**Comment by Raigi** [comment]
> > . Raigi reports seeing this with this branch and on meta-updater master but none of the release branches.
> 
> So, can I try the master branch of meta-updater to reproduce it?

Yes, I saw the same issue there

**Comment by mikesul-vd** [comment]
> > > . Raigi reports seeing this with this branch and on meta-updater master but none of the release branches.
> > 
> > 
> > So, can I try the master branch of meta-updater to reproduce it?
> 
> Yes, I saw the same issue there
> I am on this meta-updater commit in master: `32b5e19b768530f21a0f17db73fb49562d024618`

Just did the following on `32b5e19b768530f21a0f17db73fb49562d024618`
`repo sync`
`bitbake primary-image`
`bitbake secondary-image`
-- no any primary nor secondary specific parameters are defined in local.conf -- 
`../meta-updater/scripts/run-qemu-ota --no-gui --secondary-network secondary-image`
`../meta-updater/scripts/run-qemu-ota --no-gui --secondary-network primary-image`

The result is successful
`
Mar 09 10:05:08 qemux86-64 aktualizr[225]: Adding Secondary with ECU serial: aad737dafb40f018a65eaa810a04f596a7c02ffaa0a711532d1aaaca78546238 with hardware ID: qemux86-64-sndry
Mar 09 10:05:09 qemux86-64 aktualizr[225]: Provisioned successfully on Device Gateway.
Mar 09 10:05:09 qemux86-64 aktualizr[225]: ECUs have been successfully registered with the server.
Mar 09 10:05:09 qemux86-64 aktualizr[225]: Primary ECU serial: dba4d3cbbdf8820b7b87b01c7bc9297bfcf030fc47d0cd40b6aaec603ba163d2 with hardware ID: qemux86-64
Mar 09 10:05:09 qemux86-64 aktualizr[225]: Device ID: flaky-biersuppe-956

`

**Comment by patrickvacek** [comment]
@mike-sul It seems that if you start the Primary first your chances of success are much lower.

**Comment by Raigi** [comment]
> @mike-sul It seems that if you start the Primary first your chances of success are much lower.

Good catch. I've been starting primary first, then secondary, same for OTF.

**Comment by mikesul-vd** [comment]
> > @mike-sul It seems that if you start the Primary first your chances of success are much lower.
> 
> @patrickvacek Good catch. I've been starting primary first, then secondary, same for OTF.

Just tried a few times and it still works

`
Mar 09 10:50:20 qemux86-64 aktualizr[227]: Found IP secondaries config: (wait_port: 9040 timeout_s: 240)
Mar 09 10:50:20 qemux86-64 aktualizr[227]:    found IP secondary config: (addr: 192.168.254.2:9050)
Mar 09 10:50:20 qemux86-64 aktualizr[227]: Creating IP secondaries...
Mar 09 10:50:20 qemux86-64 aktualizr[227]: Connecting to and getting info about IP Secondary: 192.168.254.2:9050...
Mar 09 10:50:23 qemux86-64 aktualizr[227]: Failed to connect to a secondary: No route to host
Mar 09 10:50:23 qemux86-64 aktualizr[227]: Waiting for connection from 1 secondaries...
Mar 09 10:50:57 qemux86-64 aktualizr[227]: Accepted connection from a secondary: (192.168.254.2:9050)
Mar 09 10:50:57 qemux86-64 aktualizr[227]: Got info on IP Secondary: hw-ID: qemux86-64-sndry serial: ce06baa87de48aeadba8a6086696051a738451e23291303789a5aaa8d990562b
Mar 09 10:50:57 qemux86-64 aktualizr[227]: Adding Secondary with ECU serial: ce06baa87de48aeadba8a6086696051a738451e23291303789a5aaa8d990562b with hardware ID: qemux86-64-sndry
Mar 09 10:50:57 qemux86-64 aktualizr[227]: Provisioned successfully on Device Gateway.

`

**Comment by mikesul-vd** [comment]
@Raigi Could you please add the following info here

1. zero level logs of both Primary and Secondary
2. toml config of both Primary and Secondary
3. the json config of Secondaries on Primary
4. output of `ip a` for both Primary and Secondary

**Comment by Raigi** [comment]
> @Raigi Could you please add the following info here
> 
>     1. zero level logs of both Primary and Secondary
[primary_0.txt](https://github.com/advancedtelematic/meta-updater/files/4306287/primary_0.txt)
[secondary_0.txt](https://github.com/advancedtelematic/meta-updater/files/4306289/secondary_0.txt)

>     2. toml config of both Primary and Secondary
Which toml you mean exactly?
>     3. the json config of Secondaries on Primary
"primary-config-1.0-r0 do_install: SOTA_SECONDARY_CONFIG hasn't been specified in the local config, generate a default one" - where does this get saved to?
>     4. output of `ip a` for both Primary and Secondary

primary
`root@qemux86-64:/sysroot/home/root# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether ca:fe:c9:70:70:77 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
       valid_lft 85914sec preferred_lft 85914sec
    inet6 fec0::c8fe:c9ff:fe70:7077/64 scope site dynamic 
       valid_lft 86187sec preferred_lft 14187sec
    inet6 fe80::c8fe:c9ff:fe70:7077/64 scope link 
       valid_lft forever preferred_lft forever
3: enp0s5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether ca:fe:05:06:89:84 brd ff:ff:ff:ff:ff:ff
    inet 192.168.254.1/24 brd 192.168.254.255 scope global enp0s5
       valid_lft forever preferred_lft forever
    inet6 fe80::c8fe:5ff:fe06:8984/64 scope link 
       valid_lft forever preferred_lft forever
4: sit0@NONE: <NOARP> mtu 1480 qdisc noop qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
`
secondary
`root@qemux86-64:/sysroot/home/root# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: enp0s3: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether ca:fe:6b:3e:10:a9 brd ff:ff:ff:ff:ff:ff
    inet 10.0.2.15/24 brd 10.0.2.255 scope global dynamic enp0s3
       valid_lft 85922sec preferred_lft 85922sec
    inet6 fec0::c8fe:6bff:fe3e:10a9/64 scope site dynamic 
       valid_lft 86234sec preferred_lft 14234sec
    inet6 fe80::c8fe:6bff:fe3e:10a9/64 scope link 
       valid_lft forever preferred_lft forever
3: enp0s5: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast qlen 1000
    link/ether ca:fe:53:e2:ef:44 brd ff:ff:ff:ff:ff:ff
    inet 192.168.254.2/24 brd 192.168.254.255 scope global enp0s5
       valid_lft forever preferred_lft forever
    inet6 fe80::c8fe:53ff:fee2:ef44/64 scope link 
       valid_lft forever preferred_lft forever
4: sit0@NONE: <NOARP> mtu 1480 qdisc noop qlen 1000
    link/sit 0.0.0.0 brd 0.0.0.0
`

**Comment by mikesul-vd** [comment]
The following error message on Primary means that something is listening on 9040 port on Primary
`Creating IP secondaries...
Failed to initialize a secondary: bind: Address already in use`

**Comment by Raigi** [comment]
Without changing anything, my devices started to provision. It takes quite a while though.
`-- Logs begin at Mon 2020-03-09 12:44:17 UTC, end at Mon 2020-03-09 12:45:58 UTC. --
Mar 09 12:44:18 qemux86-64 systemd[1]: Started Aktualizr SOTA Client.
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Aktualizr version 2020.2-106-g9f2cc5d7 starting
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Reading config: "/usr/lib/sota/conf.d/20-sota-shared-cred.toml"
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Reading config: "/usr/lib/sota/conf.d/30-secondary-config.toml"
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Reading config: "/usr/lib/sota/conf.d/35-enable-auto-reboot.toml"
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Reading config: "/usr/lib/sota/conf.d/40-hardware-id.toml"
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Bootstrap empty SQL storage
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Bootstraping DB to version 23
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Could not find primary ecu serial, defaulting to empty serial: no more rows available
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Could not find hardware_id for serial : no more rows available
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Could not find primary ecu serial, set to lazy init mode
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Found IP secondaries config: (wait_port: 9040 timeout_s: 240)
Mar 09 12:44:18 qemux86-64 aktualizr[239]:    found IP secondary config: (addr: 192.168.254.2:9050)
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Creating IP secondaries...
Mar 09 12:44:18 qemux86-64 aktualizr[239]: Connecting to and getting info about IP Secondary: 192.168.254.2:9050...
Mar 09 12:45:44 qemux86-64 aktualizr[239]: Failed to connect to a secondary: Connection refused
Mar 09 12:45:44 qemux86-64 aktualizr[239]: Waiting for connection from 1 secondaries...
Mar 09 12:45:44 qemux86-64 aktualizr[239]: Accepted connection from a secondary: (192.168.254.2:9050)
Mar 09 12:45:44 qemux86-64 aktualizr[239]: Got info on IP Secondary: hw-ID: qemux86-64-sndry serial: 025161204b0d6994aae4af1e13e04387c845cae2e44d300e87e148567acb3615
Mar 09 12:45:44 qemux86-64 aktualizr[239]: Adding Secondary with ECU serial: 025161204b0d6994aae4af1e13e04387c845cae2e44d300e87e148567acb3615 with hardware ID: qemux86-64-sndry
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Provisioned successfully on Device Gateway
Mar 09 12:45:45 qemux86-64 aktualizr[239]: ECUs have been successfully registered to the server.
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Primary ECU serial: d90c36ac1a2185c17610eade0307115bbddbb75e667a1afe6103e52d4e007080 with hardware ID: qemux86-64
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Device ID: exotic-punschkrapfen-318
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Device Gateway URL: https://f442fc7c-bd84-444e-b1ee-6a634f02612a.device-gateway.ota.api.here.com:443
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Certificate subject: CN=exotic-punschkrapfen-318, OU=Devices
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Certificate issuer: CN=Device Registration
Mar 09 12:45:45 qemux86-64 aktualizr[239]: Certificate valid from: Mar  9 12:45:45 2020 GMT until: Nov  3 13:17:00 2029 GMT
Mar 09 12:45:47 qemux86-64 aktualizr[239]: got SendDeviceDataComplete event
`

**Comment by mikesul-vd** [comment]
> Without changing anything, my devices started to provision. It takes quite a while though.
> [journalctl_aktualizr.txt](https://github.com/advancedtelematic/meta-updater/files/4306629/journalctl_aktualizr.txt)
> 
> Next device creation took veery long.
> [journalct_aktualizr_long.txt](https://github.com/advancedtelematic/meta-updater/files/4306656/journalct_aktualizr_long.txt)

According to the "long" logs it looks like it has to do with networking in general, I have a few theories but it would be great to learn how to reproduce it. 

@patrickvacek can you reproduce it on bagger?

**Comment by Raigi** [comment]
> It appears that Raigi's problems are not dependent upon this PR, so I vote to merge this now and work on solving her problems separately.

This one not, but now there are new problems. :D

Having `IMAGE_INSTALL_append = " aktualizr-auto-reboot "` in local.conf does not seem to have effect on secondary.

**Comment by patrickvacek** [comment]
> @patrickvacek can you reproduce it on bagger?

Yes, that is where I was seeing it this morning.

**Comment by lbonn** [comment]
That's "interesting" that the primary also reboots. It also happened to me during a test today but not the last time I've tried last week...

**Comment by lbonn** [comment]
Will update after https://github.com/advancedtelematic/aktualizr/pull/1594 is (hopefully) merged

**Comment by lbonn** [comment]
This time, only the secondary should reboot.

**Comment by Raigi** [comment]
> This time, only the secondary should reboot.

Just to confirm that devices got provisioned, 
installing an update on secondary rebooted only secondary.

**Comment by mike-sul** [comment]
Does it still take lots of time to provision a device?

**Comment by Raigi** [comment]
> Does it still take lots of time to provision a device?

yes, it does

**Comment by patrickvacek** [comment]
@mike-sul I think these connection problems might also be connected to the failures in `test_ip_secondary` I'm seeing in ptest (see https://main.gitlab.in.here.com/olp/edge/ota/connect/client/meta-updater/-/jobs/1913879 for example). I'd like to see if the basic retries or whatever we do to improve this make a difference for that as well.

**Comment by mike-sul** [comment]
> @mike-sul I think these connection problems might also be connected to the failures in `test_ip_secondary` I'm seeing in ptest (see https://main.gitlab.in.here.com/olp/edge/ota/connect/client/meta-updater/-/jobs/1913879 for example). I'd like to see if the basic retries or whatever we do to improve this make a difference for that as well.

It looks like the issue here https://main.gitlab.in.here.com/olp/edge/ota/connect/client/meta-updater/-/jobs/1913879 is different from the one reported in this PR. In this PR we have the issue with the initial discovery/connection of Primary and Secondary during a device provisioning phase  while in that test Primary cannot get a manifest from Secondary that has been updated and is supposed to be rebooted (or reboot emulated in this case).

**Comment by patrickvacek** [comment]
> It looks like the issue here https://main.gitlab.in.here.com/olp/edge/ota/connect/client/meta-updater/-/jobs/1913879 is different from the one reported in this PR. In this PR we have the issue with the initial discovery/connection of Primary and Secondary during a device provisioning phase while in that test Primary cannot get a manifest from Secondary that has been updated and is supposed to be rebooted (or reboot emulated in this case).

Agreed, but I'm wondering if the problem in that CI run is due to similar connectivity problems after rebooting the Secondary. It could be something with the reboot process instead, though.

**Comment by mike-sul** [comment]
> > It looks like the issue here https://main.gitlab.in.here.com/olp/edge/ota/connect/client/meta-updater/-/jobs/1913879 is different from the one reported in this PR. In this PR we have the issue with the initial discovery/connection of Primary and Secondary during a device provisioning phase while in that test Primary cannot get a manifest from Secondary that has been updated and is supposed to be rebooted (or reboot emulated in this case).
> 
> Agreed, but I'm wondering if the problem in that CI run is due to similar connectivity problems after rebooting the Secondary. It could be something with the reboot process instead, though.

Yeah, I think something has changed in the networking or maybe Laurent actually fixed a bug in the IP secondary's logic that tries to connect to Primary at its startup. In the case of localhost/127.0.0.1 connect() call should return almost immediately if nothing listens on the destination port if it's not a localhost then IP Secondary gets stuck at this stage (TCP connection at TCP/IP stack level) for some times (around > 20 sec). In the test we just run Secondary process and do not wait until IP Secondary is ready to server requests instead we just run Aktualizr process and the secondary actually connects to Primary but the following getManifest() call from Primary to Secondary fails which means that Secondary was not ready to serve/listen socket at the call time...
Anyway, I think, I know how we can fix/improve the test but in general, as I stated a few times the current approach of Primary and Secondary discovery is error-prone...

**Pull request merged**